### PR TITLE
Allow circular dependencies between classes

### PR DESCRIPTION
**Pull request by AdamRzepka**
Allow code like this:

```rust
struct A {}
impl A {
    fn a(b: &B){}
}

struct B {}
impl B {
    fn b(a: &A){}
}
```

### CODE REVIEW

File: macroslib/src/cpp/rust_str.h [hunk 1]
```diff
@@ -16,7 +16,8 @@ struct RustStrView {
     bool empty() const noexcept { return this->len == 0; }
     std::string to_std_string() const { return std::string{ data, len }; }
 #endif
-#if __cplusplus > 201402L
+#if __cplusplus > 201402L && (!defined(__GNUC__) || defined(__clang__))
+// GCC5 reports `__cplusplus > 201402L`, but does not support `std::string_view`.
```
Comment (Dushistov) [inline macroslib/src/cpp/rust_str.h]: Looks strange, Why code and comment do not match?

According to comment the code should be

```
#if __cplusplus > 201402L &&  (!defined(__GNUC__) || __GNUC__ > 5)
```

but instead of `__GNUC__ > 5` it check `__clang__` for some reason?

And looks like this completely unrelated issue, may be it is possible to move this change into separate commit?

Comment (AdamRzepka) [reply macroslib/src/cpp/rust_str.h]: Clang also defines `__GNUC__`, so, if you need code specific for GCC, but not for clang, you need to test `defined(__GNUC__) && !defined(__clang__)`.
The condition is reversed here, becouse the code should be included if we are NOT in GCC.

I'll rebase it into separate commits.

Comment (Dushistov) [reply macroslib/src/cpp/rust_str.h]: > Clang also defines __GNUC__

Ok, but what about gcc with version 6-8?

should it be:
```
#if __cplusplus > 201402L &&  (!defined(__GNUC__) || defined(__clang__) || __GNUC__ > 5)
```

Comment (AdamRzepka) [reply macroslib/src/cpp/rust_str.h]: @Dushistov 
I've added more comprehensive test if `string_view` is supported and splited PR into two commits.

File: macroslib/src/java_jni/mod.rs [hunk 2]
```diff
@@ -89,16 +89,11 @@ impl ForeignMethodSignature for JniForeignMethodSignature {
 }
 
 impl LanguageGenerator for JavaConfig {
-    fn generate(
```
Comment (Dushistov) [inline macroslib/src/java_jni/mod.rs]: I added test case for your PR: b3f19f67151d7e977811900c642f8a8cc2420cf7 

and it **fails** for Java case:

> Can not find such rust type '& B' Can not find conversation from 'jlong' to '& B'

The simple approach I suppose to force it to work correctly:
-  it should parse each classes, but skip methods, and deal only with constructors, and then re-parse again with methods.

But of course if constructors accept class that not yet defined then this is fatal error,
the more robust and complex solution is
- parse only constructors, deals only with return values from constructors to maker familiar with the way how to create classes, and then repeat the same thing without ignoring method and input parameters of constructors

or may be 
- allow forward declarations, like:
```Rust
foreigner_class!(class A {
    self_type A;
    constructor A::default() -> A;  
});

foreigner_class!(class B {
    self_type B;
    constructor B::default() -> B;
    static_method B::b(a: &A);
});

foreigner_class!(class A {
    self_type A;
    constructor A::default() -> A;
    static_method A::a(b: &B);
});

```

Comment (AdamRzepka) [reply macroslib/src/java_jni/mod.rs]: Oh, I didn't notice registration code in `generate_rust_code` function. I've moved it into `register_class`, just like I did with C++ and now the test works.

Let me know if this solution is OK and if I should squash all changes into two commits, like previously.

File: macroslib/src/cpp/rust_str.h [hunk 3]
```diff
@@ -16,7 +16,9 @@ struct RustStrView {
     bool empty() const noexcept { return this->len == 0; }
     std::string to_std_string() const { return std::string{ data, len }; }
 #endif
-#if __cplusplus > 201402L
```
Comment (Dushistov) [inline macroslib/src/cpp/rust_str.h]: The main check disappears 

> __cplusplus > 201402L

so if compile code with `-std=c++11` or `-std=c++14` it fails.

File: macroslib/src/cpp/mod.rs [hunk 4]
```diff
@@ -517,21 +564,12 @@ public:
     let mut gen_code = Vec::new();
 
     let (this_type_for_method, code_box_this) =
-        if let (Some(this_type), Some(constructor_ret_type)) = (
+        if let (Some(this_type), Some(_constructor_ret_type)) = (
```
Comment (Dushistov) [inline macroslib/src/cpp/mod.rs]: If there is no need for `_constructor_ret_type` why bother to take it from expression?

File: macroslib/src/java_jni/rust_code.rs [hunk 5]
```diff
@@ -50,7 +50,7 @@ pub(in crate::java_jni) fn generate_rust_code(
 
     let mut gen_code = Vec::<TokenStream>::new();
     let (this_type_for_method, code_box_this) =
-        if let (Some(this_type), Some(constructor_ret_type)) = (
+        if let (Some(this_type), Some(_constructor_ret_type)) = (
```
Comment (Dushistov) [inline macroslib/src/java_jni/rust_code.rs]: If there is no need for _constructor_ret_type why bother to take it from expression?

File: macroslib/src/cpp/mod.rs (hunk 4 above)
Comment (AdamRzepka) [reply macroslib/src/cpp/mod.rs]: The code mimics old behaviour: only do `add_type` for `this_type_for_method` and `constructor_ret_type`, and generate `impl SwigForeignClass`, if both `this_type_for_method` and `constructor_ret_type` are present.

Now, the type registration and code generation is separated, but the behaviour is the same: if `constructor_ret_type` is `None`, code for `impl SwigForeignClass` won't be generated.

That being said, the relationship between `self_type`, `this_type_for_method`, `constructor_ret_type` and `SwigForeignClass` is a bit vague for me, which is why I decided to just recreate old behaviour, without digging into it.

File: macroslib/src/cpp/rust_str.h (hunk 3 above)
Comment (AdamRzepka) [reply macroslib/src/cpp/rust_str.h]: Right, I'll fix it

Comment (AdamRzepka) [reply macroslib/src/cpp/rust_str.h]: Meantime I've found a better way.

From the C++17 standard
```
    1 The following macro names shall be defined by the implementation:

    __cplusplus The integer literal 201703L.
```

The problem with gcc was, it defined intermediate values between 201402L and 201703L. For example in gcc 5.4 `__cplusplus` equals 201500L when run with `-std=c++17` (which, I suppose, indicates that c++17 is partially supported).

So, testing `#if __cplusplus >= 201703L` is consistent with standard and should work for gcc, clang and MSVC.

**Comment by Dushistov** [comment]
@AdamRzepka 

Thank you for your work.

I will replace your checks in rust_str.h to
> __cplusplus >= 201703L

feel free to submit new PR if it breaks something.
